packages/@magic-sdk/provider/src/modules/user.ts

```diff
@@ -47,19 +51,59 @@ export class UserModule extends BaseModule {
   }
 
   public isLoggedIn() {
-    const requestPayload = createJsonRpcRequestPayload(
-      this.sdk.testMode ? MagicPayloadMethod.IsLoggedInTestMode : MagicPayloadMethod.IsLoggedIn,
-    );
-    return this.request<boolean>(requestPayload);
+    return createPromiEvent<boolean, any>(async (resolve, reject) => {
+      try {
+        let cachedIsLoggedIn = false;
+        if (this.sdk.useStorageCache) {
+          cachedIsLoggedIn = (await getItem(this.localForageIsLoggedInKey)) === 'true';
+
+          // if isLoggedIn is true on storage, optimistically resolve with true
+          // if it is false, we use `usr.isLoggedIn` as the source of truth.
+          if (cachedIsLoggedIn) {
+            resolve(true);
+          }
+        }
+
+        const requestPayload = createJsonRpcRequestPayload(
+          this.sdk.testMode ? MagicPayloadMethod.IsLoggedInTestMode : MagicPayloadMethod.IsLoggedIn,
+        );
+        const isLoggedInResponse = await this.request<boolean>(requestPayload);
+        if (this.sdk.useStorageCache) {
+          if (!isLoggedInResponse) {
+            removeItem(this.localForageIsLoggedInKey);
+            if (cachedIsLoggedIn) {
+              this.emitUserLoggedOut(true);
+            }
+          } else {
+            setItem(this.localForageIsLoggedInKey, true);
+          }
+        }
+        resolve(isLoggedInResponse);
```

Comment:
I would suggest flattening the if condition, for better readability
```suggestion
        if (this.sdk.useStorageCache) {
          if (isLoggedInResponse) {
                      setItem(this.localForageIsLoggedInKey, true);
          } else {
                      removeItem(this.localForageIsLoggedInKey);
          }
          
            if (cachedIsLoggedIn && !isLoggedInResponse) {
              this.emitUserLoggedOut(true);
            }
        }
        resolve(isLoggedInResponse);
```